## **1. Collecting The Data:**

Machine learning starts with data, which serves as the foundation for training models. Data can come from various sources, such as:
<br>- Historical records (e.g., stock prices, sales data).
<br>- Images, audio, or video.
<br>- Real-time data such as: sensor or sensors readings.
<br>- etc.
<br>The quality and quantity of data directly impact the performance of the machine learning model.

## **2. Preprocess The Data:**

### **a. Understanding The Data by Descriptive Statistics:**

We will use the "House Sales in King County, USA" dataset from Kaggle.
<br>Kaggle dataset link:
<br>https://www.kaggle.com/datasets/harlfoxem/housesalesprediction

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
from IPython.display import display
import numpy as np
import os

In [2]:
CURRENT_DIR = os.getcwd()
PARENT_DIR = os.path.dirname(CURRENT_DIR)
GET_CSV_FILE = os.path.join(PARENT_DIR, "03.Pandas/kc_house_data _rev.csv")

df_houses = pd.read_csv(GET_CSV_FILE)

Examining the raw data can reveal valuable insights, which may help you pre-process the data, making your machine learning models perform better.

In [3]:
# View the first 10 rows in the data:

with pd.option_context('display.max_columns', None):
    display(df_houses.head(10))

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,20141013T000000,221900.0,3,1.00,1180.0,5650.0,1.0,0.0,0.0,3.0,7,1180.0,0.0,1955.0,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000.0,3,2.25,2570.0,7242.0,2.0,0.0,0.0,3.0,7,2170.0,400.0,1951.0,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000.0,2,1.00,770.0,10000.0,1.0,0.0,0.0,3.0,6,770.0,0.0,1933.0,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000.0,4,3.00,1960.0,5000.0,1.0,0.0,0.0,5.0,7,1050.0,910.0,1965.0,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000.0,3,2.00,1680.0,8080.0,1.0,0.0,0.0,3.0,8,1680.0,0.0,1987.0,0,98074,47.6168,-122.045,1800,7503
5,7237550310,20140512T000000,1230000.0,4,4.50,5420.0,101930.0,1.0,0.0,0.0,3.0,11,3890.0,1530.0,2001.0,0,98053,47.6561,-122.005,4760,101930
6,1321400060,20140627T000000,257500.0,3,2.25,1715.0,6819.0,2.0,0.0,0.0,3.0,7,1715.0,0.0,1995.0,0,98003,47.3097,-122.327,2238,6819
7,2008000270,20150115T000000,291850.0,3,1.50,1060.0,9711.0,1.0,0.0,0.0,3.0,7,1060.0,0.0,1963.0,0,98198,47.4095,-122.315,1650,9711
8,2414600126,20150415T000000,229500.0,3,1.00,1780.0,7470.0,1.0,0.0,0.0,3.0,7,1050.0,730.0,1960.0,0,98146,47.5123,-122.337,1780,8113
9,3793500160,20150312T000000,323000.0,3,2.50,1890.0,6560.0,2.0,0.0,0.0,3.0,7,1890.0,0.0,NaN,0,98038,47.3684,-122.031,2390,7570


You should check the data shape, taking into account that:
- Too many rows may take too long to train your models, for example, if you are training your model on your pc, and you are not using cloud computing web services, such as Amazon Web Services, Microsoft Azure, Google Cloud, or IBM Cloud.
- Insufficient data to train your models, leading to inaccurate results.
- The curse of dimensionality arises when working with high-dimensional data that has too many columns (features). As the number of dimensions increases, data points may lead to:
  * Data sparsity: This makes the data difficult to analyze, visualize, and extract meaningful patterns.
  * Overfitting risk: With too many features relative to samples, models can memorize noise instead of learning generalizable patterns.
  * Computational cost: Algorithms become slower and more resource-intensive as dimensionality increases.

In [4]:
print(df_houses.shape)

(21613, 21)


As shown above, our dataset has 21613 rows and 21 columns.

Now we will check the dataset information

In [5]:
df_info = df_houses.info()

print(df_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21605 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21610 non-null  float64
 5   sqft_living    21607 non-null  float64
 6   sqft_lot       21608 non-null  float64
 7   floors         21610 non-null  float64
 8   waterfront     21609 non-null  float64
 9   view           21608 non-null  float64
 10  condition      21610 non-null  float64
 11  grade          21613 non-null  int64  
 12  sqft_above     21606 non-null  float64
 13  sqft_basement  21611 non-null  float64
 14  yr_built       21606 non-null  float64
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

#### **a.1 Dataset's descriptive statistics:**
They are used to review observations from the results:

In [6]:
# TO control the number of decimal places displayed for floating-point numbers.
# We will use the "set_option" function to round the data to 2 decimal places.

pd.set_option('display.float_format', '{:.2f}'.format)


houses_statistics = df_houses.describe()

display(houses_statistics)

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
count,21613.00,21605.00,21613.00,21610.00,21607.00,21608.00,21610.00,21609.00,21608.00,21610.00,21613.00,21606.00,21611.00,21606.00,21613.00,21613.00,21613.00,21613.00,21613.00,21613.00
mean,4580301520.86,539779.77,3.37,2.11,2079.80,15097.85,1.49,0.01,0.23,3.41,7.66,1788.49,291.48,1971.00,84.40,98077.94,47.56,-122.21,1986.55,12768.46
std,2876565571.31,366576.42,0.93,0.77,918.46,41403.65,0.54,0.09,0.77,0.65,1.18,828.15,442.58,29.38,401.68,53.51,0.14,0.14,685.39,27304.18
min,1000102.00,75000.00,0.00,0.00,290.00,520.00,1.00,0.00,0.00,1.00,1.00,290.00,0.00,1900.00,0.00,98001.00,47.16,-122.52,399.00,651.00
25%,2123049194.00,321500.00,3.00,1.75,1426.00,5040.00,1.00,0.00,0.00,3.00,7.00,1190.00,0.00,1951.00,0.00,98033.00,47.47,-122.33,1490.00,5100.00
50%,3904930410.00,450000.00,3.00,2.25,1910.00,7617.00,1.50,0.00,0.00,3.00,7.00,1560.00,0.00,1975.00,0.00,98065.00,47.57,-122.23,1840.00,7620.00
75%,7308900445.00,645000.00,4.00,2.50,2550.00,10687.25,2.00,0.00,0.00,4.00,8.00,2210.00,560.00,1997.00,0.00,98118.00,47.68,-122.12,2360.00,10083.00
max,9900000190.00,7700000.00,33.00,8.00,13540.00,1651359.00,3.50,1.00,4.00,5.00,13.00,9410.00,4820.00,2015.00,2015.00,98199.00,47.78,-121.31,6210.00,871200.00


The above shows only the numeric columns and ignores non-numeric data
<br>If you need categorical summary stats, use "describe(include=['O'])" for object (string) types.

Here’s how you can go beyond just reading the numbers:
1. Detect Missing Data
- Compare count with total rows → highlights columns with missing values.
2. Spot Outliers
- Large gap between min/max and quartiles suggests extreme values.
- High std relative to mean indicates wide variability.
3. Understand Distribution
- Compare mean vs. 50% (median):
- If they differ significantly → skewed distribution.
- Quartiles show clustering of values (e.g., income data often skewed).
4. Categorical Insights
- Unique count reveals diversity of categories.
- Top/freq shows the dominant category.
    - Both points will be discussed in the categorical data part.
5. Feature Engineering Clues
- Columns with little variance (low std, one dominant category) may be poor predictors.
- Skewed numeric features may need transformation (log, normalization).

#### **a.2 Categorical Data:**
Categorical data refers to the data that is grouped into distinct categories rather than measured numerically. It is often used to classify observations based on labels or names rather than numerical values. There are two main types:
- Nominal Data: Categories have no inherent order (e.g., colors, gender, country names).
- Ordinal Data: Categories have a meaningful order (e.g., in our dataset, grade, condition, and they are numeric data, they may be non-numeric data such as (A > B > C > .....).

In [7]:
# You can check the size of each attribute in this Ordinal Data:
grade_count = df_houses.groupby("grade").size()

print(grade_count)

grade
1        1
3        3
4       29
5      242
6     2038
7     8981
8     6068
9     2615
10    1134
11     399
12      90
13      13
dtype: int64


In [8]:
grade_count = df_houses.groupby("condition").size()

print(grade_count)

condition
1.00       30
2.00      172
3.00    14030
4.00     5677
5.00     1701
dtype: int64


In [9]:
# To change the "condition" column values from numbers to strings,
# use the "replace" function:
df_houses["condition_cat"] = df_houses["condition"].replace({1: "E", 2: "D", 3: "C", 4: "B", 5: "A"})

df_houses["condition_cat"].unique()

array(['C', 'A', 'B', 'E', 'D', nan], dtype=object)

In [10]:
# As for the grade column, it represents the overall construction
# and design quality of the house, as assessed by King County’s
# grading system:
# - Low Quality: 1–4
# - Average Quality: 5–7
# - Good Quality: 8–10
# - Luxury: 11–13

# OR
# To change the "grade" column values from numbers to strings,
# use the "map" function:
mapping = {
    1: "Low", 2: "Low", 3: "Low", 4: "Low",
    5: "Average", 6: "Average", 7: "Average",
    8: "Good", 9: "Good", 10: "Good",
    11: "Luxury", 12: "Luxury", 13: "Luxury",
}
df_houses["grade_cat"] = df_houses["grade"].map(mapping)

df_houses["grade_cat"].unique()

array(['Average', 'Good', 'Luxury', 'Low'], dtype=object)

In [11]:
houses_cat = df_houses[['condition_cat', 'grade_cat']]
houses_cat.describe(include=['O'])

,condition_cat,grade_cat
count,21610,21613
unique,5,4
top,C,Average
freq,14030,11261


- count → total non-null entries
- unique → number of distinct categories
- top → the most common category
- freq → how many times that top category appears

This table is useful because it gives you a quick diagnostic snapshot of your categorical data:
- **Data completeness** → "count" shows how many non-null values exist per column. You can spot missing data patterns immediately.
  
- **Category richness** → "unique" tells you how many distinct categories exist. This helps you determine whether a feature is diverse (with many categories) or narrow (with few categories).

- **Dominant category** → "top "and "freq" highlight the most common value and how often it appears. This reveals an imbalance or skew in categorical features.

- **Feature usefulness** → If one category dominates (high freq relative to count), the feature may have low predictive power and could be dropped or transformed.

- **Data validation** → You can catch anomalies, like placeholder values (“Unknown”, “NA”) showing up as the top category, which signals data quality issues.

- **Encoding strategy** → Knowing the top category helps decide between one-hot encoding, target encoding (mean encoding), or grouping rare categories. For example, if freq is high, you might collapse rare categories into “Other.”

#### **a.3 Correlations Between Attributes:**
In machine learning, correlation between attributes is significant because it helps in understanding relationships between features, improving model performance, and avoiding redundancy.

Correlation is the relationship between two variables and how they may or may not change together.
<br>Positive correlations between attributes indicate that both increase together, while negative ones mean one decreases as the other increases.
<br>A correlation of -1 or 1 shows a strong correlation between the attributes. Whereas a value of 0 shows no correlation at all.

Some machine learning algorithms, such as linear and logistic regression, may perform poorly if the correlation between attributes is strong.
<br>Thus, it is a good idea to review all of the pairwise correlations of the attributes in your dataset.

The most common method for calculating correlation is Pearson’s Correlation Coefficient, which assumes a normal distribution of the attributes involved.

We will use the "corr" method, which is only applicable to a numeric dataset.
<br>If the dataset has non-numeric attributes, remove them when using the "corr" method.
<br> In our dataset, some columns shouldn't be used with the "corr" method:
- The "date" is non-numeric.
- The "id " column is just an ID for each house.
- The categorical data, the "grade" and "condtion" columns.

In [12]:
# Dropping unnecessary columns:
data_drop_non_num = df_houses.drop(["date", "id", "grade", "condition", "condition_cat", "grade_cat"], axis=1)

data_corr = data_drop_non_num.corr(method="pearson")
display(data_corr)

,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
price,1.00,0.31,0.53,0.70,0.09,0.26,0.27,0.40,0.61,0.32,0.05,0.12,-0.05,0.31,0.02,0.58,0.08
bedrooms,0.31,1.00,0.52,0.58,0.03,0.18,-0.01,0.08,0.48,0.30,0.15,0.02,-0.15,-0.01,0.13,0.39,0.03
bathrooms,0.53,0.52,1.00,0.75,0.09,0.50,0.06,0.19,0.69,0.28,0.51,0.05,-0.20,0.02,0.22,0.57,0.09
sqft_living,0.70,0.58,0.75,1.00,0.17,0.35,0.10,0.28,0.88,0.43,0.32,0.05,-0.20,0.05,0.24,0.76,0.18
sqft_lot,0.09,0.03,0.09,0.17,1.00,-0.01,0.02,0.07,0.18,0.02,0.05,0.01,-0.13,-0.09,0.23,0.14,0.72
floors,0.26,0.18,0.50,0.35,-0.01,1.00,0.02,0.03,0.52,-0.25,0.49,0.01,-0.06,0.05,0.13,0.28,-0.01
waterfront,0.27,-0.01,0.06,0.10,0.02,0.02,1.00,0.40,0.07,0.08,-0.03,0.09,0.03,-0.01,-0.04,0.09,0.03
view,0.40,0.08,0.19,0.28,0.07,0.03,0.40,1.00,0.17,0.28,-0.05,0.10,0.08,0.01,-0.08,0.28,0.07
sqft_above,0.61,0.48,0.69,0.88,0.18,0.52,0.07,0.17,1.00,-0.05,0.42,0.02,-0.26,-0.00,0.34,0.73,0.19
sqft_basement,0.32,0.30,0.28,0.43,0.02,-0.25,0.08,0.28,-0.05,1.00,-0.13,0.07,0.07,0.11,-0.14,0.20,0.02


Correlation helps us with:
- Feature Selection: Highly correlated features can introduce redundancy, so removing one can simplify the model without losing valuable information.
- Multicollinearity Detection: In regression models, strong correlations between independent variables can distort predictions, making it harder to interpret coefficients.
- Predictive Power: Some features may have strong correlations with the target variable, making them more useful for predictions.

We can use the "spearman" correlation with oridnal data:

In [13]:
data_corr = df_houses.loc[:, ["grade", "condition"]].corr(method="spearman")
display(data_corr)

,grade,condition
grade,1.00,-0.17
condition,-0.17,1.00


We will discuss correlation later using the Correlation Matrix Plot.